File: backend/src/main/kotlin/top/bearingwall/asya/util/JwtUtil.kt
package top.bearingwall.asya.util

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.security.Keys
import io.jsonwebtoken.security.WeakKeyException
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.Date
import java.util.UUID
import javax.crypto.SecretKey

object JwtUtil {
    private const val DEFAULT_SECRET = "asya-backend-jwt-secret-key-change-me-to-at-least-32-chars"
    private const val MIN_KEY_BYTES = 32
    private const val ACCESS_EXPIRATION_SECONDS = 3600L
    private const val REFRESH_EXPIRATION_SECONDS = 30L * 24 * 60 * 60
    private const val CLAIM_TYPE = "type"
    private const val CLAIM_AUTH_VERSION = "ver"

    private val key: SecretKey by lazy { buildKey() }

    private fun buildKey(): SecretKey {
        val configured = System.getenv("JWT_SECRET")
            ?: System.getProperty("jwt.secret")
            ?: DEFAULT_SECRET
        val keyBytes = configured.toByteArray(StandardCharsets.UTF_8)
        require(keyBytes.size >= MIN_KEY_BYTES) {
            "JWT secret must be at least $MIN_KEY_BYTES bytes; configure JWT_SECRET env or -Djwt.secret"
        }
        return try {
            Keys.hmacShaKeyFor(keyBytes)
        } catch (e: WeakKeyException) {
            throw IllegalStateException("JWT secret too weak: ${e.message}", e)
        }
    }

    fun generateAccessToken(subject: String, claims: Map<String, Any> = emptyMap(), authVersion: Int): String {
        return generateToken(
            subject = subject,
            claims = claims + mapOf(
                CLAIM_TYPE to TokenType.ACCESS.claimValue,
                CLAIM_AUTH_VERSION to authVersion
            ),
            expirationSeconds = ACCESS_EXPIRATION_SECONDS
        )
    }

    fun generateRefreshToken(subject: String, authVersion: Int): String {
        return generateToken(
            subject = subject,
            claims = mapOf(
                CLAIM_TYPE to TokenType.REFRESH.claimValue,
                CLAIM_AUTH_VERSION to authVersion
            ),
            expirationSeconds = REFRESH_EXPIRATION_SECONDS
        )
    }

    private fun generateToken(subject: String, claims: Map<String, Any>, expirationSeconds: Long): String {
        val now = Instant.now()
        return Jwts.builder()
            .subject(subject)
            .claims(claims)
            .id(UUID.randomUUID().toString())
            .issuedAt(Date.from(now))
            .expiration(Date.from(now.plusSeconds(expirationSeconds)))
            .signWith(key, Jwts.SIG.HS256)
            .compact()
    }

    // Parse JWT and return subject and claims. Throws if invalid/expired.
    fun parseToken(token: String): ParsedToken {
        val claims = Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .payload
        return ParsedToken(
            subject = claims.subject,
            claims = claims
        )
    }

    fun requireTokenType(parsedToken: ParsedToken, expectedType: TokenType) {
        val actualType = parsedToken.claims[CLAIM_TYPE]?.toString()
        if (actualType != expectedType.claimValue) {
            throw JwtException("Token类型无效")
        }
    }

    fun getAuthVersion(parsedToken: ParsedToken): Int {
        val rawVersion = parsedToken.claims[CLAIM_AUTH_VERSION]
            ?: throw JwtException("Token缺少认证版本")
        return when (rawVersion) {
            is Number -> rawVersion.toInt()
            is String -> rawVersion.toIntOrNull() ?: throw JwtException("Token认证版本无效")
            else -> throw JwtException("Token认证版本无效")
        }
    }
}

data class ParsedToken(
    val subject: String,
    val claims: io.jsonwebtoken.Claims
)

enum class TokenType(val claimValue: String) {
    ACCESS("access"),
    REFRESH("refresh")
}

File: backend-dotnet/src/AsyaMun.Api/Auth/JwtUtil.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using AsyaMun.Api.Errors;
using AsyaMun.Api.Models;
using Microsoft.IdentityModel.Tokens;

namespace AsyaMun.Api.Auth;

public static class JwtUtil
{
    private const string DefaultSecret = "asya-backend-jwt-secret-key-change-me-to-at-least-32-chars";
    private const int MinKeyBytes = 32;
    private const long AccessExpirationSeconds = 3600L;
    private const long RefreshExpirationSeconds = 30L * 24 * 60 * 60;
    private const string ClaimType = "type";
    private const string ClaimAuthVersion = "ver";

    private static readonly Lazy<SymmetricSecurityKey> Key = new(BuildKey);

    private static SymmetricSecurityKey BuildKey()
    {
        var configured = Environment.GetEnvironmentVariable("JWT_SECRET");
        if (string.IsNullOrWhiteSpace(configured))
        {
            configured = DefaultSecret;
        }

        var keyBytes = Encoding.UTF8.GetBytes(configured);
        if (keyBytes.Length < MinKeyBytes)
        {
            throw new InvalidOperationException(
                $"JWT secret must be at least {MinKeyBytes} bytes; configure JWT_SECRET env or -Djwt.secret");
        }

        return new SymmetricSecurityKey(keyBytes);
    }

    public static SymmetricSecurityKey GetSigningKey() => Key.Value;

    public static string GenerateAccessToken(string subject, int authVersion, string? name = null, UserRole? role = null)
    {
        var claims = new Dictionary<string, object>
        {
            [ClaimAuthVersion] = authVersion,
            [ClaimType] = TokenType.Access.ClaimValue()
        };

        if (name != null)
        {
            claims["name"] = name;
        }

        if (role != null)
        {
            claims["role"] = role.Value.ToString();
        }

        return GenerateToken(subject, claims, TimeSpan.FromSeconds(AccessExpirationSeconds));
    }

    public static string GenerateRefreshToken(string subject, int authVersion)
    {
        return GenerateToken(
            subject,
            new Dictionary<string, object>
            {
                [ClaimType] = TokenType.Refresh.ClaimValue(),
                [ClaimAuthVersion] = authVersion
            },
            TimeSpan.FromSeconds(RefreshExpirationSeconds));
    }

    private static string GenerateToken(string subject, Dictionary<string, object> claims, TimeSpan lifetime)
    {
        var now = DateTime.UtcNow;
        var securityClaims = new List<Claim> { new(JwtRegisteredClaimNames.Sub, subject) };
        securityClaims.AddRange(claims.Select(kv => new Claim(kv.Key, kv.Value.ToString() ?? string.Empty)));
        securityClaims.Add(new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString("N")));

        var token = new JwtSecurityToken(
            claims: securityClaims,
            notBefore: now,
            expires: now.Add(lifetime),
            signingCredentials: new SigningCredentials(Key.Value, SecurityAlgorithms.HmacSha256));

        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    public static ParsedToken ParseToken(string token)
    {
        var handler = new JwtSecurityTokenHandler();
        var validationParameters = new TokenValidationParameters
        {
            IssuerSigningKey = Key.Value,
            ValidateIssuerSigningKey = true,
            ValidateIssuer = false,
            ValidateAudience = false,
            ValidateLifetime = true,
            ClockSkew = TimeSpan.Zero,
            ValidAlgorithms = new[] { SecurityAlgorithms.HmacSha256 }
        };

        var principal = handler.ValidateToken(token, validationParameters, out var validatedToken);
        var claims = principal.Claims.ToDictionary(c => c.Type, c => (object?)c.Value);
        var subject = principal.FindFirst(JwtRegisteredClaimNames.Sub)?.Value
            ?? principal.FindFirst(ClaimTypes.NameIdentifier)?.Value
            ?? throw new InvalidOperationException("Token缺少subject");

        return new ParsedToken(subject, claims);
    }

    public static void RequireTokenType(ParsedToken parsedToken, TokenType expectedType)
    {
        var actualType = parsedToken.Claims.TryGetValue(ClaimType, out var value) ? value?.ToString() : null;
        if (!string.Equals(actualType, expectedType.ClaimValue(), StringComparison.Ordinal))
        {
            throw new TokenInvalidException("Token类型无效");
        }
    }

    public static int GetAuthVersion(ParsedToken parsedToken)
    {
        if (!parsedToken.Claims.TryGetValue(ClaimAuthVersion, out var raw) || raw == null)
        {
            throw new TokenInvalidException("Token缺少认证版本");
        }

        return raw switch
        {
            int i => i,
            long l => (int)l,
            string s when int.TryParse(s, out var v) => v,
            _ => throw new TokenInvalidException("Token认证版本无效")
        };
    }
}

public record ParsedToken(string Subject, Dictionary<string, object?> Claims);

public enum TokenType
{
    Access,
    Refresh
}

public static class TokenTypeExtensions
{
    private const string AccessValue = "access";
    private const string RefreshValue = "refresh";

    public static string ClaimValue(this TokenType type)
    {
        return type switch
        {
            TokenType.Access => AccessValue,
            TokenType.Refresh => RefreshValue,
            _ => throw new ArgumentOutOfRangeException(nameof(type))
        };
    }
}